_source/_posts/2021-11-16-k8s-security-best-practices.md

```diff
@@ -0,0 +1,148 @@
+
+When someone has access to a Kubernetes cluster, they can access the k8s API server and SSH into the cluster nodes themselves. Ideally, this should be limited as much as possible. Disable SSH access to non-admin users. Secure the API server using OIDC and RBAC, as we saw earlier so that only authenticated users with sufficient roles can access the API.
+
+## 6. Secure container images
+
+Securing the container images that run on the cluster is as important as securing the cluster itself. A malicious image running on a cluster could wreak havoc. Follow these best practices for container image security.
+
+- Do not run containers as root as this would give the container unlimited access to the host. Always run the containers using a non-root user
+- Enable container image scanning in your CI/CD phase to catch known vulnerabilities using tools like [clair](https://github.com/quay/clair) or [Anchore](https://github.com/anchore/anchore-engine)
+- Use minimal up to date official base images and remove all unwanted dependencies, packages, and debugging tools from the image as it will make it more secure and lightweight
+- Prevent loading unwanted kernel modules in the containers. These can be restricted using rules in `/etc/modprobe.d/kubernetes-blacklist.conf` of the node or by uninstalling the unwanted modules from the node.
+- Use [official verified images](https://docs.docker.com/docker-hub/official_images/) for popular software. Use a trusted registry for non-official images and always verify the image publisher
+- Use [Docker Bench for Security](https://github.com/docker/docker-bench-security) to audit your container images
+- Use [Pod security policies](https://kubernetes.io/docs/concepts/policy/pod-security-policy/) to limit a container's access to the host further
+
+You can read more about it in our "[Container Security: A Developer Guide](https://developer.okta.com/blog/2019/07/18/container-security-a-developer-guide)".
+
+## 7. Control traffic between pods and clusters
+
+Generally, pods within the same cluster will be able to communicate with each other, and if you have multiple clusters in the same network, there may be traffic between them as well. Do not let this all open, as it could lead to a compromised cluster when another in the network is affected. Use [Kubernetes network policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/) to control traffic between pods and clusters and allow only necessary traffic.
+
+## 8. Use namespaces to isolate workloads
+
+Do not run all your workloads in a single namespace. Isolating workloads in different namespaces based on business needs will make it more secure and easier to manage with RBAC. This way, you can fine-tune RBAC even further to let users access only what they need to see. You can also use Kubernetes network policies to isolate traffic between namespaces where applicable.
```

Comment:
```suggestion
Do not run all your workloads in a single namespace. Isolating workloads in different namespaces based on business needs is more secure and easier to manage with RBAC. This way, you can fine-tune RBAC even further to let users access only what they need to see. You can also use Kubernetes network policies to isolate traffic between namespaces where applicable.
```
clarifying, otherwise it is unclear what the "it" refers to.